agent/lib_php_amqplib.c

```diff
@@ -0,0 +1,758 @@
+ * before and after the function call so we can just maintain pointers into the
+ * struct.  With PHP 7.x, without doing special handling, we don't have access
+ * to the values afterwards.  Sometimes nr_php_arg_get is used as that DUPs the
+ * zval which then later needs to be freed with nr_php_arg_release.  In this
+ * case, we don't need to go through the extra trouble of duplicating a ZVAL
+ * when we don't need to duplicate anything if there is no valid value.  We
+ * check for a valid value, and if we want to store it, we'll strdup it. So
+ * instead of doing multiple zval dups all of the time, we do some strdups some
+ * of the time.
+ */
+#if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO /* PHP8.0+ */
+#define ENSURE_PERSISTENCE(x) x
+#define UNDO_PERSISTENCE(x)
+#else
+#define ENSURE_PERSISTENCE(x) nr_strdup(x)
+#define UNDO_PERSISTENCE(x) nr_free(x);
+#endif
+
+/*
+ * See here for supported Amazon MQ for RabbitMQ engine versions
+ * https://docs.aws.amazon.com/amazon-mq/latest/developer-guide/rabbitmq-version-management.html
+ * For instance:
+ * As of Feb 2025, 3.13 (recommended)
+ *
+ * See here for latest RabbitMQ Server https://www.rabbitmq.com/docs/download
+ * For instance:
+ * As of Feb 2025, the latest release of RabbitMQ Server is 4.0.5.
+ *
+ * https://www.rabbitmq.com/tutorials/tutorial-one-php
+ * Installing RabbitMQ
+ *
+ * While the RabbitMQ tutorial for using with the dockerized RabbitMQ setup
+ * correctly and loads the PhpAmqpLib\\Channel\\AMQPChannel class in time for
+ * the agent to wrap the instrumented functions, there are AWS MQ_BROKER
+ * specific but valid scenarios where the PhpAmqpLib\\Channel\\AMQPChannel class
+ * file does not explicitly load or does not load in time, and the instrumented
+ * functions are NEVER wrapped regardless of how many times they are called in
+ * one txn.  Specifically, this centered around the very slight but impactful
+ * differences when using managing the AWS MQ Broker connect vs
+ * causes an explicit load of the AMQPChannel class/file and
+ * PhpAmqpLib\Connection\AMQPSSLConnection which does NOT cause an explicit load
+ * of the AMQPChannelclass/file. The following method is thus the only way to
+ * ensure the class is loaded in time for the functions to be wrapped.
+ *
+ */
+
+/*
+ * Purpose : Ensures the php-amqplib instrumentation gets wrapped.
+ *
+ * Params  : None
+ *
+ * Returns : None
+ */
+static void nr_php_amqplib_ensure_class() {
+  zval retval_zpd;
+  int result = FAILURE;
+//amber
+return;
+
+  result = zend_eval_string("class_exists('PhpAmqpLib\\Channel\\AMQPChannel');",
+                            &retval_zpd, "Get nr_php_amqplib_class_exists");
+  /*
+   * We don't need to check anything else at this point. If this fails, there's
+   * nothing else we can do anyway.
+   */
+
+  zval_dtor(&retval_zpd);
+}
+
+/*
+ * Version detection will be called directly from PhpAmqpLib\\Package::VERSION
```

Comment:
What does it mean to be called from a class constant? `PhpAmqpLib\\Package::VERSION` is a class constant: https://github.com/php-amqplib/php-amqplib/blob/master/PhpAmqpLib/Package.php#L8 so how can it call anything.

Reply:
[4aa2522](https://github.com/newrelic/newrelic-php-agent/pull/1009/commits/4aa25223a683824afbcfabe17905fd6bdd24314d)